src/test/java/htsjdk/samtools/BAMSBIIndexerTest.java

```diff
@@ -0,0 +1,157 @@
+            final SBIIndex index1 = fromBAMFile(BAM_FILE, g);
+            final SBIIndex index2 = fromSAMRecords(BAM_FILE, g);
+            Assert.assertEquals(index1, index2);
+            Assert.assertEquals(index1.dataFileLength(), bamFileSize);
+            Assert.assertEquals(index2.dataFileLength(), bamFileSize);
+        }
+    }
+
+    private SBIIndex fromBAMFile(final File bamFile, final long granularity) throws IOException {
+        final ByteArrayOutputStream out = new ByteArrayOutputStream();
+        BAMSBIIndexer.createIndex(new SeekableFileStream(bamFile), out, granularity);
+        return SBIIndex.load(new ByteArrayInputStream(out.toByteArray()));
+    }
+
+    private SBIIndex fromSAMRecords(final File bamFile, final long granularity) {
+        final ByteArrayOutputStream out = new ByteArrayOutputStream();
+        final SBIIndexWriter indexWriter = new SBIIndexWriter(out, granularity);
+        final BAMFileReader bamFileReader = bamFileReader(bamFile);
+        final CloseableIterator<SAMRecord> iterator = bamFileReader.getIterator();
+        while (iterator.hasNext()) {
+            processAlignment(indexWriter, iterator.next());
+        }
+        indexWriter.finish(bamFileReader.getVirtualFilePointer(), bamFile.length());
+        return SBIIndex.load(new ByteArrayInputStream(out.toByteArray()));
+    }
+
+    public void processAlignment(final SBIIndexWriter indexWriter, final SAMRecord rec) {
+        final SAMFileSource source = rec.getFileSource();
+        if (source == null) {
+            throw new SAMException("No source (virtual file offsets); needed for indexing on BAM Record " + rec);
+        }
+        final BAMFileSpan filePointer = (BAMFileSpan) source.getFilePointer();
+        indexWriter.processRecord(filePointer.getFirstOffset());
+    }
+
+    private BAMFileReader bamFileReader(final File bamFile) {
+        final SamReader samReader = SamReaderFactory.makeDefault()
+                .validationStringency(ValidationStringency.SILENT)
+                .enable(SamReaderFactory.Option.INCLUDE_SOURCE_IN_RECORDS).open(bamFile);
+        return (BAMFileReader) ((SamReader.PrimitiveSamReaderToSamReaderAdapter) samReader).underlyingReader();
+    }
+
+    private SAMRecord getReadAtOffset(final File bamFile, final long virtualOffset) {
+        final Chunk chunk = new Chunk(virtualOffset, BlockCompressedFilePointerUtil.makeFilePointer(bamFile.length()));
+        try (CloseableIterator<SAMRecord> iterator = bamFileReader(bamFile).getIterator(new BAMFileSpan(chunk))) {
+            Assert.assertTrue(iterator.hasNext());
+            return iterator.next();
+        }
+    }
+
+    private List<SAMRecord> getReadsInChunk(final File bamFile, final Chunk chunk) {
+        try (CloseableIterator<SAMRecord> iterator = bamFileReader(bamFile).getIterator(new BAMFileSpan(chunk))) {
+            return Iterables.slurp(iterator);
+        }
+    }
+}
```

Comment:
this is actually a comment for the bam but I can't comment on it. 

Could you use `src/test/resources/htsjdk/samtools/cram/CEUTrio.HiSeq.WGS.b37.NA12878.20.first.8000.bam` instead? It is already included in the test data.

Reply:
Yes, that works fine. Deleted the new BAM that I added.